File: apps/api/app/services/fetchers/linkedin.py
"""LinkedIn fetcher — post metrics via Marketing / Community Management API."""

import logging

import httpx

from app.services.fetchers.base import BaseFetcher, NormalizedMetrics

logger = logging.getLogger(__name__)


class LinkedInFetcher(BaseFetcher):
    """Fetch LinkedIn post engagement via shareStatistics and socialActions.

    Requires OAuth 2.0 token with r_organization_social or r_member_social scope.
    """

    platform_id = "linkedin"

    async def fetch(self, token: str, post_id: str) -> NormalizedMetrics | None:
        headers = {
            "Authorization": f"Bearer {token}",
            "X-Restli-Protocol-Version": "2.0.0",
            "LinkedIn-Version": "202401",
        }

        try:
            impressions = 0
            clicks = 0
            likes = 0
            comments = 0
            shares = 0

            async with httpx.AsyncClient(timeout=15) as client:
                # Primary: organizational share statistics
                stats_resp = await client.get(
                    "https://api.linkedin.com/rest/organizationalEntityShareStatistics",
                    params={
                        "q": "organizationalEntity",
                        "shares": f"List({post_id})",
                    },
                    headers=headers,
                )

                if stats_resp.status_code == 200:
                    elements = stats_resp.json().get("elements", [])
                    if elements:
                        stats = elements[0].get("totalShareStatistics", {})
                        impressions = stats.get("impressionCount", 0)
                        clicks = stats.get("clickCount", 0)
                        likes = stats.get("likeCount", 0)
                        comments = stats.get("commentCount", 0)
                        shares = stats.get("shareCount", 0)

                if not impressions and not likes:
                    # Fallback: social actions for basic engagement counts
                    actions_resp = await client.get(
                        f"https://api.linkedin.com/rest/socialActions/{post_id}",
                        headers=headers,
                    )
                    if actions_resp.status_code == 200:
                        data = actions_resp.json()
                        likes = data.get("likesSummary", {}).get("totalLikes", 0)
                        comments = data.get("commentsSummary", {}).get("totalFirstLevelComments", 0)

            engagements = likes + comments + shares + clicks
            engagement_rate = round(engagements / max(impressions, 1), 6)

            return NormalizedMetrics(
                impressions=impressions,
                engagements=engagements,
                engagement_rate=engagement_rate,
                saves_bookmarks=0,
                shares_reposts=shares,
                comments=comments,
                clicks=clicks,
                follows_gained=0,
                platform_specific={
                    "likes": likes,
                    "comments": comments,
                    "shares": shares,
                    "clicks": clicks,
                    "impressions": impressions,
                },
            )

        except Exception as exc:
            logger.warning("Failed to fetch LinkedIn metrics for post %s: %s", post_id, exc)
            return None

File: functions/src/lib/analytics/fetchers/linkedin.ts
/**
 * LinkedIn metrics fetcher — Marketing API socialActions + shares.
 */

import type { MetricsFetcher, FetchMetricsOptions, PlatformMetrics } from "./base.js";
import { handleApiResponse } from "./base.js";

interface LinkedInShareStats {
  totalShareStatistics?: {
    shareCount?: number;
    clickCount?: number;
    likeCount?: number;
    commentCount?: number;
    impressionCount?: number;
    engagement?: number;
  };
}

export const linkedinFetcher: MetricsFetcher = {
  platformId: "linkedin",

  async fetchMetrics(options: FetchMetricsOptions): Promise<PlatformMetrics> {
    const { platformPostId, accessToken } = options;

    // LinkedIn uses URN-based post IDs: urn:li:share:{id} or urn:li:ugcPost:{id}
    const urn = platformPostId.startsWith("urn:") ? platformPostId : `urn:li:share:${platformPostId}`;
    const encodedUrn = encodeURIComponent(urn);

    const resp = await fetch(
      `https://api.linkedin.com/v2/organizationalEntityShareStatistics?q=organizationalEntity&shares=List(${encodedUrn})`,
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "X-Restli-Protocol-Version": "2.0.0",
        },
      }
    );

    // Fallback: try socialActions endpoint if share stats fail
    if (!resp.ok) {
      return await fetchViaSocialActions(platformPostId, accessToken, urn);
    }

    const json = await resp.json() as { elements?: LinkedInShareStats[] };
    const stats = json.elements?.[0]?.totalShareStatistics;

    if (!stats) {
      return await fetchViaSocialActions(platformPostId, accessToken, urn);
    }

    return {
      impressions: stats.impressionCount ?? 0,
      engagements: (stats.likeCount ?? 0) + (stats.commentCount ?? 0) + (stats.shareCount ?? 0) + (stats.clickCount ?? 0),
      saves: 0,
      shares: stats.shareCount ?? 0,
      clicks: stats.clickCount ?? 0,
      follows: 0,
      comments: stats.commentCount ?? 0,
    };
  },
};

async function fetchViaSocialActions(
  platformPostId: string,
  accessToken: string,
  urn: string
): Promise<PlatformMetrics> {
  const encodedUrn = encodeURIComponent(urn);
  const resp = await fetch(
    `https://api.linkedin.com/v2/socialActions/${encodedUrn}`,
    {
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "X-Restli-Protocol-Version": "2.0.0",
      },
    }
  );

  handleApiResponse("linkedin", resp, platformPostId);

  const json = await resp.json() as {
    likesSummary?: { totalLikes?: number };
    commentsSummary?: { totalFirstLevelComments?: number };
  };

  const likes = json.likesSummary?.totalLikes ?? 0;
  const comments = json.commentsSummary?.totalFirstLevelComments ?? 0;

  return {
    impressions: 0,
    engagements: likes + comments,
    saves: 0,
    shares: 0,
    clicks: 0,
    follows: 0,
    comments,
  };
}
